### DISCUSSION TITLE
pyOCD and Holtek HT32 devboard and devices

### DISCUSSION THREAD

**Discussion by acontini** [pyocd/pyOCD#1073] [Q&A]
Hi all, newbie here.
I don't want to open a bug report because maybe it's my pyOCD configuration or maybe it is a bug in the CMSIS pack that I'm using. First of all, thank you for developing such a useful tool.

I currently have a Holtek ESK32-30505 starter kit and I cannot program the target chip.
The output of the command _pyocd flash --verbose -t ht32f52253 Test3.elf_  is:
0001790:INFO:board:Target type is ht32f52253
0001880:INFO:dap:DP IDR = 0x0bc11477 (v1 MINDP rev0)
0001899:INFO:ap:AHB-AP#0 IDR = 0x04770031 (AHB-AP var3 rev0)
0001928:INFO:rom_table:AHB-AP#0 Class 0x1 ROM table #0 @ 0xe00ff000 (designer=43b part=4c0)
0001943:INFO:rom_table:[0]<e000e000:SCS-M0+ class=14 designer=43b part=008>
0001952:INFO:rom_table:[1]<e0001000:DWT-M0+ class=14 designer=43b part=00a>
0001959:INFO:rom_table:[2]<e0002000:BPU class=14 designer=43b part=00b>
0001962:INFO:cortex_m:CPU core #0 is Cortex-M0+ r0p1
0001969:INFO:dwt:2 hardware watchpoints
0001973:INFO:fpb:4 hardware breakpoints, 0 literal comparators
0001982:INFO:loader:Erased 0 bytes (0 sectors), programmed 0 bytes (0 pages), skipped 0 bytes (0 pages) at 0.00 kB/s

Looking at the last line, it seems that the file is not copied into the target. I've tried to google the error with no luck.
Do you think it is something related to a bug in the firmware file of the pack?
Can you please suggest more tests that I can do to debug this error?
Any other ideas to please help me? 

This board includes an e-link32 CMSIS-DAP compatible probe and a HT32F52253 target chip.
In Windows, with uVision 5 and the correct ARM licence, it works perfectly.
I was going to give up as I'm currently using Linux Debian 10, and no software seemed to support it, until I've read that pyOCD can recognise CMSIS-DAP probes and I gave it a try.
The output of the _pyocd list_ is:
```
  #   Probe                     Unique ID  
-------------------------------------------
  0   HOLTEK Holtek CMSIS-DAP   0900001A   

```
Great!
I have erased the chip successfully. It had a blink LEDs example flashed in it, and after the erase both LEDs are off.
The output is:
_pyocd erase --verbose --chip -t ht32f52253_
0001801:INFO:board:Target type is ht32f52253
0001890:INFO:dap:DP IDR = 0x0bc11477 (v1 MINDP rev0)
0001908:INFO:ap:AHB-AP#0 IDR = 0x04770031 (AHB-AP var3 rev0)
0001935:INFO:rom_table:AHB-AP#0 Class 0x1 ROM table #0 @ 0xe00ff000 (designer=43b part=4c0)
0001954:INFO:rom_table:[0]<e000e000:SCS-M0+ class=14 designer=43b part=008>
0001961:INFO:rom_table:[1]<e0001000:DWT-M0+ class=14 designer=43b part=00a>
0001967:INFO:rom_table:[2]<e0002000:BPU class=14 designer=43b part=00b>
0001972:INFO:cortex_m:CPU core #0 is Cortex-M0+ r0p1
0001978:INFO:dwt:2 hardware watchpoints
0001983:INFO:fpb:4 hardware breakpoints, 0 literal comparators
0001989:INFO:eraser:Erasing chip...
0002133:INFO:eraser:Done

**Comment by flit** [comment]
Hi @acontini! It could be that your ELF doesn't have any programmable sections. This would happen if it's linked to run out of RAM instead of flash. Since erasing worked, it seems likely flashing would work too. (Usually it either works ok or not at all.)

Assuming you're using the [GNU Arm Embedded toolchain](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm), could you try running `arm-non-eabi-objdump -h Test3.elf` and copy the log here? This will list the sections with addresses and attributes.

Another option is to try converting it to a binary image: `arm-none-eabi-objcopy -O binary Test3.elf Test3.bin`

**Reply by acontini** [reply]
Hi @flit, thanks for the hint.

For a newbie is quite difficult to get oriented around so many tools!
Please let me know where do you think the problem could be (i.e. linker, compiler, etc.) and I'll have a better look.

The result of objdump -h is:
```
Test3.elf:     file format elf32-littlearm

Sections:
Idx Name          Size      VMA       LMA       File off  Algn
  0 .isr_vector   00000000  00000000  00000000  00010000  2**0
                  CONTENTS
  1 .text         00000000  00000000  00000000  00010000  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
  2 .rodata       00000000  00000000  00000000  00010000  2**0
                  CONTENTS, ALLOC, LOAD, DATA
  3 .data         00000000  20000000  20000000  00010000  2**0
                  CONTENTS, ALLOC, LOAD, DATA
  4 .bss          00000000  20000000  20000000  00000000  2**0
                  ALLOC
```

**Reply by acontini** [reply]
Hi @flit, at the end it was a sum of different issues with the startup code, the linker flags and the clock prescaler settings. I have now resolved and the board LEDs are back blinking as before. Thank you for the support.

**Reply by flit** [reply]
Glad you got it working! And yeah, there's a lot to learn. 😄
